int32[] ids
int32[] values